Add Tasmota IRHVAC integration

hristo-atanasov opened the pull request:
This component allows you to control hundreds of ACs via Tasmota based IR Transceiver, using the IRHVAC command in Tasmota

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adding a platform, that will help to lots of users to control hundreds of ACs out of the box, using Tasmota based IR Transceivers 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
climate:
  - platform: tasmota_irhvac # Required
    name: "Daewoo IRHvac" #The name of Your device
    command_topic: "cmnd/kitchenMultisensor/irhvac" #the topic to send IRHVAC command to Your Tasmota IR Transceiver
    state_topic: "tele/kitchenMultisensor/RESULT" #The toppic that receives IR codes
    temperature_sensor: sensor.kitchen_temperature # The entity ID of the temperature sensor remove if no such
    protocol: "ELECTRA_AC" #Change to your protocol or leave it if it is the same
    min_temp: 16 #optional - default 16 int value - Your AC's min temp - remove if it is the same
    max_temp: 32 #optional - default 32 int value - Your AC's max temp - remove if it is the same
    target_temp: 26 #optional - default 26 int value - enter Your initial target temp
    initial_operation_mode: "off" # optional - default "off" string value (one of the "supported_modes") - remove if it is the same
    away_temp: 24 #optional - default 24 int value - remove if it is the same
    precision: 1 #optional - default 1 int or float value. Can be set to 1, 0.5 or 0.1 - remove if it is the same
    supported_modes:
      # Leave uncommented only the modes that Your AC support
      - "heat"
      - "cool"
      - "dry"
      - "fan_only"
      - "auto"
      - "off" #Turns the AC off - Should be quoted
      # Some devices have "auto" and "fan_only" chaned
      # If following two lines are uncommented, "auto" and "fan" shoud be commented
      #- auto_fan_only #if remote shows fan but tasmota says auto
      #- fan_only_auto #if remote shows auto but tasmota says fan
    supported_fan_speeds:
      #Uncoment only the supported by Your AC fan speeds
      # Some devices say max,but it is high, and auto which is max
      # If you uncomment the following two, you have to comment high and max
      # - auto_max #woud become max
      # - max_high #would become high
      #- on
      #- off
      #- low
      - "medium"
      - "high"
      #- middle
      #- focus
      #- diffuse
      - "min"
      - "max"
      #- auto
    supported_swing_list:
      - "off"
      - "vertical" #up to down - comment this i Your AC does not support vertical swinging
      # - horizontal # Left to right - uncomment if your AC supports horizontal swinging
      # - both # Top to Bottom - uncomment if your AC supports both horizontal and vertical swinging
    default_quiet_mode: "Off" #optional - default "Off" string value - remove if it is the same
    default_turbo_mode: "Off" #optional - default "Off" string value - remove if it is the same
    default_econo_mode: "Off" #optional - default "Off" string value - remove if it is the same
    hvac_model: "-1" #optional - default "1" string value
    celsius_mode: "On" # If remote sends the temp in celsius optional - default "On", string value - remove if it is the same
    default_light_mode: "Off" #optional - default "Off" string value - remove if it is the same
    default_filter_mode: "Off" #optional - default "Off" string value - remove if it is the same
    default_clean_mode: "Off" #optional - default "Off" string value - remove if it is the same
    default_beep_mode: "Off" #optional - default "Off" string value - remove if it is the same
    default_sleep_mode: "-1" #optional - default "-1" string value - remove if it is the same
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
https://github.com/hristo-atanasov/Tasmota-IRHVAC/blob/master/README.md

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

hristo-atanasov committed:
Tasmota IRHVAC initial upload

This component allows you to control hundreds of ACs via Tasmota based IR Transceiver, using the IRHVAC command in Tasmota

homeassistant commented:
Hi @hristo-atanasov, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#31422).

Once you do that we will be able to review and accept this pull request.

Thanks!

springstan commented:
@hristo-atanasov you will need to do a couple of things to solve the CI errors:

1. Fix your typo by renaming `maniest.json` to `manifest.json`
2. Run `codespell` to fix any spelling mistakes.
3. Format your code by running `black --fast homeassistant tests` inside of your development environment.

Hope this helps you to get started ✌

hristo-atanasov committed:
Fixed typo from "maniest.json" to "manifest.json"

springstan commented on homeassistant/components/tasmota_irhvac/manifest.json:
```diff
@@ -0,0 +1,11 @@
+{
+    "domain": "tasmota_irhvac",
+    "name": "Tasmota Irhvac",
+    "documentation": "https://github.com/hristo-atanasov/Tasmota-IRHVAC",
```
The documentation needs to be hosted on [home-assistant.io](https://www.home-assistant.io) therefore you will need to open a PR [here](https://github.com/home-assistant/home-assistant.io) to add this documentation :)
```suggestion
    "documentation": "https://www.home-assistant.io/integrations/tasmota_irhvac",
```

hristo-atanasov replied:
Thanks for your help, man .. This is the first time I'making PR here and I'm not so much in the situation .. :)

springstan replied:
No worries 😄 you should take a look at the [development checklist](https://developers.home-assistant.io/docs/en/development_checklist.html) which lists requirements for new integrations and other contributions toi help you solve common problems.

hristo-atanasov replied:
I have one question .. I tried to add myself as code owner, but my username contains "-" in the middle and Visual Studio Code shows it as error .. Is it really error or I can add myself? Or just leave it blank? I don't need to be "famous" .. :)

springstan replied:
>  Is it really error or I can add myself?

No, that should not be an error, so please add yourself :) Visual Studio Code should only complain if the codeowner name is not enclosed in "".

hristo-atanasov replied:
@springstan  I'm running hassio (Home Assistant) on VirtualBox .. I'm not sure how to run
"python3 -m script.gen_requirements_all", "codespell" and "black --fast homeassistant tests" ..

hristo-atanasov replied:
I made everything, but I still have these 2 checks that were not successful. What to do now?

hristo-atanasov committed:
Update manifest.json

hristo-atanasov committed:
Fixed typos and used Black

hristo-atanasov committed:
Removed commented code block

hristo-atanasov committed:
Added code owner

hristo-atanasov committed:
Fix "Missing docstring in public package"

hristo-atanasov committed:
Fix PyLint logging-not-lazy

hristo-atanasov committed:
CODEOWNERS file updated

springstan commented on homeassistant/components/tasmota_irhvac/const.py:
```diff
@@ -0,0 +1,205 @@
+"""Provides the constants needed for component."""
+
+# All activity disabled / Device is off/standby
```
Please use and import as many constants from [homeassistant/const.py](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/const.py) and [climate/const.py](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/climate/const.py) as possible.

hristo-atanasov replied:
Ok, I'load almost all constants from there ..

hristo-atanasov replied:
Done!

springstan commented on homeassistant/components/tasmota_irhvac/__init__.py:
```diff
@@ -0,0 +1 @@
+"""The climate component."""
```
```suggestion
"""The Tasmota Irhvac climate component."""
```

springstan commented on homeassistant/components/tasmota_irhvac/climate.py:
```diff
@@ -0,0 +1,748 @@
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the generic thermostat platform."""
+    name = config.get(CONF_NAME)
```
Please use `dict.get(key)` only for optional keys without default value and `dict[key]` for optional keys with default value and required keys.

hristo-atanasov replied:
Ok, will do!

hristo-atanasov replied:
Done!

springstan commented on homeassistant/components/tasmota_irhvac/climate.py:
```diff
@@ -0,0 +1,748 @@
+            swing_v = STATE_AUTO
+        # Poplate the payload
+        payload = (
+            '{"Vendor":"'
```
Use f-strings instead of string concatenation, see the [style guidelines](https://developers.home-assistant.io/docs/en/next/development_guidelines.html#use-new-style-string-formatting) for more information.

hristo-atanasov replied:
Done!

springstan commented on homeassistant/components/tasmota_irhvac/manifest.json:
```diff
@@ -0,0 +1,11 @@
+    "requirements": [],
+    "dependencies": [
+      "mqtt",
+      "sensor"
```
Why is `sensor` as a dependency?

hristo-atanasov replied:
I'm not sure if it should be dependency, but in the code, I'm getting temperature sensor data and I'm looking for state change (with a callback) of this sensor .. If it isn't needed I'll remove it but I can't decide by myself. Can You, please, help me about that?

springstan replied:
For what do you need this temperature sensor exactly?

hristo-atanasov replied:
I'm showing its value on the climate card .. This is the actual current temp in the room .. Not the target temp ..

hristo-atanasov replied:
Not sure what to do so no chagne ..

springstan commented on homeassistant/components/tasmota_irhvac/climate.py:
```diff
@@ -0,0 +1,748 @@
+    | SUPPORT_PRESET_MODE
+)
+
+DEFAULT_NAME = "IR AirConditioner"
```
I would recommend to move all of these constants to const.py and check if you can reuse any constants from the two constant files I mentioned.

hristo-atanasov replied:
Nice suggestion! I was guiding myself by looking how everything is made in other components and I haven't had that idea at all. Thank You for helping me!

springstan replied:
Of course!

hristo-atanasov replied:
Done!

springstan commented on homeassistant/components/tasmota_irhvac/manifest.json:
```diff
@@ -0,0 +1,11 @@
+      "sensor"
+    ],
+    "codeowners": ["@hristo-atanasov"]
+  }
```
Remove services.yaml since it is empty.

hristo-atanasov replied:
Ok, I will remove it.

hristo-atanasov replied:
Done!

hristo-atanasov committed:
Update homeassistant/components/tasmota_irhvac/__init__.py

Co-Authored-By: springstan <46536646+springstan@users.noreply.github.com>

hristo-atanasov committed:
Code improvements

Moved and reused contants
Removed services.yalml
Use dict.get(key) only for optional keys without default value and dict[key] for optional keys with default value and required keys.

MartinHjelmare commented on homeassistant/components/tasmota_irhvac/climate.py:
```diff
@@ -0,0 +1,695 @@
+        """(Re)Subscribe to topics."""
+
+        @callback
+        def state_message_received(msg):
```
Please extract the mqtt topic and payload parsing and subscription logic to a standalone library published on pypi.

See the mysensors integration for an example how to integrate and use the home assistant mqtt integration.

hristo-atanasov replied:
Is your proposal mandatory or just an advice?

springstan replied:
Mandatory since it is in our [development checklist](https://developers.home-assistant.io/docs/en/development_checklist.html).

hristo-atanasov replied:
Mandatory is only the part with the communication with other devices. As MQTT is the communication with other devices, it is separate part from my integration. I guess, you don't want to say that balllob and frenk are writing against their own rules, as the integrations that they are making and supporting are currently using `mqtt` in the same way. I don't see any of you stopped the `mqtt` integrations, like the `mqtt.switch` for example, or `manual_mqtt.alarm_control_panel`. I see Martin is very proud, that he has moved his mqtt requests outside of his `climate.py`, but this is his his style of codding in his situation with multiple communication protocols supported. Good job ,by the way. But I don't see any reason to accept Your advice for mandatory as more than the half of the integrations are currently written the same way, by the main developers of HA, like @balloob and @frenck  .. Please revise their integrations first, then revise mine again .. If You still think I have to create a separate class just to make one topic subscription and one topic publishing - no problem .. I can use my own integration as custom integration and I'll stop this pull request. But I'm interested in sharing it with everyone, as I believe it is useful. I think that you, guys, are understanding the mentioned above part of the check list in a different way. I'm not saying my integrations is the best in the world. I'm not even Python developer. I just decided to do the things the way that HA developers do it.

pvizeli replied:
Please follow the advice of Martin and don't tag Paulus or Franck. Mqtt is a transport protocol for what we use it.

frenck replied:
The integration and platforms you are shielding your statement with are implementing core platform components, your integration is not.

For integrations, we require logic to be abstracted into separate libraries. This has been done for a long time now and is documented. 

So I agree on the call Martin makes here since it fits.

hristo-atanasov replied:
@frenck I'm sorry I passed the ball to you. I saw You in the contributors list for these platforms, and I know You're developing for HA, so thought You created these platforms.

@pvizeli As integrations count is too big, can you point me to another integration, that has its use of mqtt outside of the main integration file? I want to see if there is a simple way of doing things. Martin's code it too complicated and best suits the use of multiple protocols. I feel I need something simpler and cleaner.

As for the other part, I feel very proud of myself for developing core platform component. :D

MartinHjelmare replied:
In basic terms, you want to pass callbacks to your library so that the library can hook into the home assistant mqtt integration. Here's the example from mysensors for that:

```py
mqtt = hass.components.mqtt

def pub_callback(topic, payload, qos, retain):
    """Call MQTT publish function."""
    mqtt.async_publish(topic, payload, qos, retain)

def sub_callback(topic, sub_cb, qos):
    """Call MQTT subscribe function."""

    @callback
    def internal_callback(msg):
        """Call callback."""
        sub_cb(msg.topic, msg.payload, msg.qos)

    hass.async_create_task(mqtt.async_subscribe(topic, internal_callback, qos))
```

This example uses the async interface so needs to be run within the event loop also in the library. If the library will not support asyncio we have to change to use the home assistant sync interface.

MartinHjelmare replied:
Here's a sync example:
```py
mqtt = hass.components.mqtt

def pub_callback(topic, payload, qos, retain):
    """Call MQTT publish function."""
    mqtt.publish(topic, payload, qos, retain)

def sub_callback(topic, sub_cb, qos):
    """Call MQTT subscribe function."""

    def internal_callback(msg):
        """Call callback."""
        sub_cb(msg.topic, msg.payload, msg.qos)

    mqtt.subscribe(topic, internal_callback, qos)
```

MartinHjelmare commented on homeassistant/components/tasmota_irhvac/climate.py:
```diff
@@ -0,0 +1,695 @@
+        await self.hass.async_add_executor_job(self.send_ir)
+        await self.async_update_ha_state()
+
+    def send_ir(self):
```
This method should be part of the library as it contains protocol specific information.

See the mysensors integration for an example how to pass the mqtt publish callback to the 3rd party library.

MartinHjelmare reviewed (changes requested):
Please extract the mqtt topic and payload subscription and publish logic and protocol specific logic to a standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#5-communication-with-devicesservices

hristo-atanasov committed:
Changing the MQTT payload from concatinated string, to JSON string  by dict serialization

hristo-atanasov committed:
Removed wildcard import

balloob commented:
i did not like your tone a single bit. You are trying to be smart (?), make odd demands and rude.

I am going to close this PR. Feel free to open a new PR once you have extracted a PyPI package and ready to interact with us in a normal manner.

Also note that code in the code base cannot be used as an argument to now make the mistake again.

hristo-atanasov commented:
@balloob How did You get my tone as "rude" from just a text .. I even put a smile emoticon in order to show that I'm not offensive.. If I was trying to be smart, believe me there is a way and place to show I'm smarter than some other people, not here. This is not a place for such thing. I think I showed my respect to Martin, as I even told that. Even excused to Frenk, that I mentioned him, also removed all the mentions, including Yours. Why every one is thinking I'm rude? Just because I mentioned You and Your code? Or because I said I followed Your code? All I said is that I used Your code as a sample for mine, because I'm not so deep in HA code like You and Frenk. You, guys are finding some not existing offence in my words. Please excuse my "rude" text. It was written with no bad words and thoughts. In this PR I was only driven by the desire to share something useful. Sorry, but English is not my native language and I can't get what You find as "rude". I guest there should be standards for "not rude" comments also. :\ Any way .. It doesn't matter, that You and (probably) everyone else here, got my words as rude, I still respect Your project and the job everyone is doing here, and I still love HA.

PS (not for people without any sense for humor):
But I also like Your last sentence, that I cannot use the current code base as an argument to make the same mistakes again! Please, excuse me again, but It just sounds funny! :D (Just a regular sense of humor, not a "rude" tone, I swear!)

Have a great week, fellas! And smile sometimes. Not every one is trying to be rude with You. Cheers! ;)
